Enhance some examples in conversions.md

RexJaeschke opened the pull request:
Tweak 3 examples to make it easier to get them through the test harness.

RexJaeschke committed:
Tweak examples to better fit test harness

RexJaeschke committed:
Update conversions.md

RexJaeschke committed:
Update conversions.md

jskeet reviewed (commented):
Hmm. Not sure about change the example here. Or maybe just move the enum to the top-level instead of it being nested?

jskeet commented on standard/conversions.md:
```diff
@@ -21,19 +21,25 @@ Some conversions in the language are defined from expressions to types, others f
->
-> // Conversion from lambda expression to delegate type
-> Func<int, int> square = x => x * x;
+> class A
```
Note: I think we're using 4-space indents, aren't we?

BillWagner commented:
closing and reopening to force a build.

RexJaeschke committed:
fixed formatting

jskeet reviewed (commented):
Again, let's leave this open until we've decided whether to create a template that has a class in so that we don't need the changes.

RexJaeschke commented:
@jskeet ready for another look by you

jskeet commented:
Will look tomorrow, thanks.

jskeet commented:
The branch is currently 4 commits behind draft-v7, so first I'm going to rebase it onto draft-v7, to make sure we don't get collisions.

jskeet commented on standard/conversions.md:
```diff
@@ -25,17 +25,23 @@ Some conversions in the language are defined from expressions to types, others f
 >
-> // The expression 0 converts implicitly to enum types
-> Color c0 = 0;
+> class A
```
The introduction of an extra class and method here still bothers me.

If we have a new template of "library code in a class" we could stick with the original code - the statements would be introducing instance fields, and the enum would be nested. A bit weird, but legal.

Any thoughts on that?

jskeet replied:
Will try this in a new commit...

jskeet replied:
Yes, this works. I'm not sure it's going to be a great long-term solution, but I think we'll want the new template anyway.

jskeet commented on standard/conversions.md:
```diff
@@ -217,29 +223,31 @@ Boxing a value of a *nullable_value_type* produces a null reference if it is the
 > {
 >     public int x, y;
->
+> 
```
I suspect this will cause a trailing whitespace warning.

jskeet replied:
(Removing.)

jskeet commented on standard/conversions.md:
```diff
@@ -217,29 +223,31 @@ Boxing a value of a *nullable_value_type* produces a null reference if it is the
-> A boxing conversion implies making a copy of the value being boxed. This is different from a conversion of a *reference_type* to type `object`, in which the value continues to reference the same instance and simply is regarded as the less derived type `object`. For example, given the declaration
+> A boxing conversion implies making a copy of the value being boxed. This is different from a conversion of a *reference_type* to type `object`, in which the value continues to reference the same instance and simply is regarded as the less derived type `object`. For example, the following
 >
 > <!-- IncompleteExample: {template:"standalone-lib", name:"BoxingConversions4", expectedErrors:["x","x"], expectedWarnings:["x","x"]} -->
```
I think this can be just "Example" now?

jskeet replied:
(I'm fixing it to be so.)

jskeet commented on standard/conversions.md:
```diff
@@ -739,44 +747,56 @@ Specifically, an anonymous function `F` is compatible with a delegate type `D`
+>         D d5 = delegate(int x) { return; };          // Ok
+>         D d6 = delegate(int x) { return x; };        // Error, return type mismatch
+>     }
+>     delegate void E(out int x);
```
I think all of this can benefit from the "code-in-class-lib" template introduced for the earlier example.

jskeet replied:
(Done - which significantly reduces the changes required for this example.)

jskeet replied:
I'd missed that the methods had been introduced as well - I've removed those, so it's declaring fields instead of locals. Seems to work! That's in the final commit.

jskeet committed:
Add new code-in-class-lib template, and amend conversions examples

All examples pass.

jskeet reviewed (commented):
I've added another commit which addresses my own comments.

Please take a look, and if you're happy with this approach of "you send a PR, I amend it" then I'll do the same for the other PR.

jskeet committed:
Tweak AnonymousFunctionsConv1 example

RexJaeschke commented:
@jskeet @BillWagner If I've understood all comments correctly, Jon has fixed all the issues he raised. If so, it's ready for your review.

jskeet commented:
Will have a look on Monday.

RexJaeschke committed:
Merge branch 'dotnet:draft-v7' into Rex-tweak-conversions-examples

RexJaeschke committed:
add core using directives

this mirrors the changes in PR #[643](https://github.com/dotnet/csharpstandard/pull/643).

jskeet commented on standard/conversions.md:
```diff
@@ -211,16 +211,17 @@ Boxing a value of a *nullable_value_type* produces a null reference if it is the
+> A boxing conversion implies making a copy of the value being boxed. This is different from a conversion of a *reference_type* to type `object`, in which the value continues to reference the same instance and simply is regarded as the less derived type `object`. For example, the following
 >
-> <!-- IncompleteExample: {template:"standalone-lib", name:"BoxingConversions4", expectedErrors:["x","x"], expectedWarnings:["x","x"]} -->
+> <!-- Example: {template:"standalone-lib", name:"BoxingConversions4"} -->
```
We could convert this to standalone-console and test the output. But we can do that later.

jskeet reviewed (approved):
I've made a slight change to one example so that it passes, and there's one thing we might want to change later on, but I'll merge now to avoid this getting stale.
